physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90

```diff
@@ -0,0 +1,7377 @@
+            endif
+            if (gam .lt. 0.12) then
+                qa100 = 0.
+            elseif (gam .gt. 1.85) then
+                qa100 = 1.0
+            else
+                qa100 = - 0.0913 + 0.7213 * gam + 0.1060 * gam ** 2 - 0.0946 * gam ** 3
+                qa100 = max (0.0, min (1., qa100))
+            endif
+            qa (k) = qa10 + (log10 (gsize / 1000.) - 1) * (qa100 - qa10)
+            qa (k) = max (0.0, min (1., qa (k)))
+        endif
+
+    enddo
+
+end subroutine cloud_fraction
+
+! =======================================================================
+! piecewise parabolic lagrangian scheme
+! this subroutine is the same as map1_q2 in fv_mapz_mod.
+! =======================================================================
+
+subroutine lagrangian_fall (ks, ke, zs, ze, zt, dp, q, precip, m1)
+
+    implicit none
+
+    ! -----------------------------------------------------------------------
+    ! input / output arguments
+    ! -----------------------------------------------------------------------
+
+    integer, intent (in) :: ks, ke
+
+    real, intent (in) :: zs
+
+    real, intent (in), dimension (ks:ke + 1) :: ze, zt
+
+    real, intent (in), dimension (ks:ke) :: dp
+
+    real, intent (inout), dimension (ks:ke) :: q
+
+    real, intent (inout) :: precip
+
+    real, intent (out), dimension (ks:ke) :: m1
+
+    ! -----------------------------------------------------------------------
+    ! local variables
+    ! -----------------------------------------------------------------------
+
+    integer :: k, k0, n, m
+
+    real :: a4 (4, ks:ke), pl, pr, delz, esl
+
+    real, parameter :: r3 = 1. / 3., r23 = 2. / 3.
+
+    real, dimension (ks:ke) :: qm, dz
+
+    ! -----------------------------------------------------------------------
+    ! density:
+    ! -----------------------------------------------------------------------
+
+    do k = ks, ke
+        dz (k) = zt (k) - zt (k + 1)
+        q (k) = q (k) * dp (k)
+        a4 (1, k) = q (k) / dz (k)
+        qm (k) = 0.
+    enddo
+
+    ! -----------------------------------------------------------------------
+    ! construct vertical profile with zt as coordinate
+    ! -----------------------------------------------------------------------
+
+    call cs_profile (a4 (1, ks), dz (ks), ke - ks + 1)
+
+    k0 = ks
+    do k = ks, ke
+        do n = k0, ke
+            if (ze (k) .le. zt (n) .and. ze (k) .ge. zt (n + 1)) then
+                pl = (zt (n) - ze (k)) / dz (n)
+                if (zt (n + 1) .le. ze (k + 1)) then
+                    ! entire new grid is within the original grid
+                    pr = (zt (n) - ze (k + 1)) / dz (n)
+                    qm (k) = a4 (2, n) + 0.5 * (a4 (4, n) + a4 (3, n) - a4 (2, n)) * (pr + pl) - &
+                        a4 (4, n) * r3 * (pr * (pr + pl) + pl ** 2)
+                    qm (k) = qm (k) * (ze (k) - ze (k + 1))
+                    k0 = n
+                    goto 555
```

Comment:
GOTOs are generally frowned upon for CCPP schemes. Consider trying to use a different control structure if possible.